Batch of revisions to locale.c

khwilliamson opened the pull request:
This is a batch of commits to locale.c that together fix bugs and improve efficiency and code readability.  A significant set of tests will be added at the conclusion of all the batches

khwilliamson committed:
Rename SETLOCALE_LOCK to LOCALE_LOCK

This mutex isn't just about the setlocale function; it applies to all
locale related critical sections.

There are no uses of SETLOCALE_LOCK in CPAN

khwilliamson committed:
locale.c: Rename to S_calculate_LC_ALL_string()

Its return is a string.  Add the suffix '_string' to emphasize that

khwilliamson committed:
locale.c: Use proper macro

querylocale_c() likely will expand to the correct thing, but it isn't
defined until higher in the stack of locale handling layers.  Instead
use the construct that is lower in the stack.

khwilliamson committed:
locale.c: Replace an assumption by a cpp check

This commit adds a preprocessor directive to check if we are running on
a POSIX 2008 system, rather than just assume we are if the earlier
possibilities weren't the case.  And it causes a compile time error if
none pan out.

khwilliamson committed:
locale.c: Don't compile without LC_ALL under POSIX 2008

The code assumes that if you have POSIX 2008, you have LC_ALL.  This
adds a compile-time error check to prevent someone from finding that out
the harder way.

khwilliamson committed:
locale.c: Move a parallel array adjacent to its mates

This is in preparation for making their parallelism more obvious

khwilliamson committed:
locale.c: Properly identify reason for placeholder in arrays

This is needed to handle the case where a locale category passed in (via
Perl_setlocale) is unknown to perl, hence is considered illegal.

The previous comments were misleading.

khwilliamson committed:
locale.c: Revamp initialization of parallel arrays

This new method, suggested by Tony Cook, uses a header file to hold
the locale categories used on this system, each as a call to a macro
with the correct parameters for that category.

When a data structure needs to use the categories, it re-#defines the
macro to generate the data it needs and #includes the header.  This
guarantees that the various parallel arrays, are in fact parallel.

This method eliminates a lot of #ifdef's in each of the data structure
initializations.

khwilliamson committed:
locale.c: Replace loop with a switch()

locale.c uses tables to specify data and actions about the different
locale categories on the system.  The categories are opaque integers
specified by libc, so they have to be mapped to indices into the tables
for our use.  Prior to this commit, that was done by a loop iterating
over each possibility until it found a match.  This commit changes that
to use a switch() statement, so that the compiler can choose what the
best way to implement things are.

khwilliamson committed:
locale.c: Move some static functions within file

This places them more conveniently.

Future commits will continue this reordering until like-level and
like-kind functions are close to each other physically in the file.

khwilliamson committed:
Change return of S_get_category_index_nowarn

This is in preparation for it being called from more than a single
place.  Instead of returning -1 on error, it returns an out-of-bounds
value 1 greater than the largest legal value.  There is a placeholder
entry in all the arrays this can be indexed into for that illegal value.
This means if it somehow gets used, we wont get a segfault, like we
would on the -1, but instead a value that will show up as wrong, but not
crash.

khwilliamson committed:
locale.c: Refactor S_get_category_index()

There is actually only one place where the locale category can be input
from code outside this file.  Now that the previous commit added
validation to that one spot, The rest of the calls should already be
validated, and if the input is bad, it means that there is a serious
error in the logic of this code, and warrants panicking.

This commit collapses two functions into one plus a macro to call it
with the right parameters.  It moves the warning into the one place
where it is useful, and panics if there is a problem otherwise.

And it adds tests for the warning.

khwilliamson committed:
locale.c: Don't compile Win32 setlocale if no locales

 If we are compiled to ignore locales, there is no need to compile the
 code that handles changing them.

khwilliamson committed:
locale.c: Fix Win32 compile if LC_(NUMERIC|MONETARY) missing

It's unlikely, but legal, to have one of these categories missing on a
platform; prior to this commit, it would generate a segfault trying to
build perl.

khwilliamson committed:
locale.c: Add STATIC to two S_ function declarations

khwilliamson committed:
locale.c: Create a longer unbroken #ifdef scope

This code is not needed when USE_LOCALE is not defined, and hence should
not be compiled.  This commit revises some #ifdef's to create a longer
stretch of functions that should thus be skipped.
S_find_locale_from_environment() is marked in embed.fnc as not to be
compiled without USE_LOCALE, but that #ifdef was missing in locale.c.
Now it is absorbed into the long scope.

Preprocessor directives indentation is adjusted as a result of this
change, and to correct previous mistakes in it.

khwilliamson committed:
Change name of private function, and when compiled

The previous short name '_warn_problematic_locale' begins with an
underscore, which is reserved for the platform's implementation.

This commit gets rid of the short name, which allows us to not need an
underscore which was used purely to avoid namespace clashes wih XS code.

The function is designed to only be called from a single macro, defined
only in core and Perl extensions.  It further doesn't get called unless
LC_CTYPE is to be used on the platform.  So restrict its definition to
that.

khwilliamson committed:
locale.c: Add some DEBUG statements

I have found these useful.

tonycoz commented on locale.c:
```diff
@@ -1345,7 +1345,18 @@ S_emulate_setlocale_i(pTHX_
+        if (PL_cur_locale_obj != new_obj) {
+            PerlIO_printf(Perl_debug_log,
+                          "(%" LINE_Tf "): PL_cur_locale_obj"
+                          " was %p, now is %p\n",
+                          line, PL_cur_locale_obj, new_obj);
```
I don't think `locale_t` is required to be a pointer type.

I've probably missed this elsewhere.  I think it can be assumed to be a scalar type (pointer or arithmetic type), since you can cast `0` to `locale_t`.

khwilliamson replied:
For example, in the man page
``locale_t newlocale(int category_mask, const char *locale,
                           locale_t base);``
it says

> The  newlocale()  function  creates a new locale object, or modifies an
        existing object, returning a reference to the new or modified object as
        the function result.

Am I wrong to think that one can always print a reference with `%p`?

khwilliamson replied:
Actually POSIX 2017, says 

> Upon successful completion, the newlocale() function shall return a handle

so I suppose that doesn't have to be a pointer; I've just never seen an implementation that isn't.  There are plenty of `%p` instances in blead for printing locale objects.
I see three possibilities:

1.  Continue using `%p` until and if it becomes a problem some day.
2. Add a Configure probe and a `LOCALE_Tf` define and convert to use it now
3. Just add the `LOCALE_Tf` define and set it to `%p`, convert to use it, and add a Configure probe if and when it becomes needed

tonycoz replied:
I went [by](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/locale.h.html):
```
The <locale.h> header shall define the locale_t type, representing a locale object.
```
which doesn't place any constraints on the type of `locale_t`, and being able to cast `0` to to `locale_t`.

I think 1 is fine.

Technically the argument of `%p` is required to be pointer to void (rather than any pointer) but I expect we break this rule generally.

tonycoz commented on locale.c:
```diff
@@ -251,225 +251,282 @@ static const char C_thousands_sep[] = "";
+#    define LC_CTYPE_MASK_                       LC_CTYPE_MASK,
+#    define LC_CTYPE_STRING_                    "LC_CTYPE",
+#      define LC_CTYPE_UPDATE_          S_new_ctype,
+#  else
```
Would this series of commits be more maintainable if it worked something like `intrpvar.h`?

It would significantly reduce the repetition: how many times is `LC_CTYPE` repeated here, including as substrings of macro names?

For example locale_table.h could be:
```
#ifdef USE_LOCALE_CTYPE
  LOCALE_TABLE_ENTRY(LC_CTYPE, LC_CTYPE_MASK, "LC_CTYPE", S_new_ctype)
#endif
#ifdef USE_LOCALE_NUMERIC
  LOCALE_TABLE_ENTRY(LC_NUMERIC, LC_NUMERIC_MASK, "LC_NUMERIC", S_new_numeric)
#endif
... and more
```
Then in locale.c:
```
#undef LOCALE_TABLE_ENTRY
#define LOCALE_TABLE_ENTRY(index, mask, name, cb) index,
STATIC const int categories[] = {
#include "locale_table.h"
  FAKE_LCALL+1
}
#undef LOCALE_TABLE_ENTRY
#defined LOCALE_TABLE_ENTRY(index, mask, name, cb) name,
STATIC const char * const category_names[] = {
#include "locale_table.h"
   "Locale category unknown to Perl; ..."
}
... etc
```
though the mask and name could be synthesized from the index parameter.

khwilliamson replied:
This sounds like a good idea.  I'll work on it.

khwilliamson committed:
locale.c: Avoid runtime strlen() calls

The lengths of these strings are known at compile time.  This commit
creates an array of those lengths to avoid calling strlen() in a loop
unnecessarily.

khwilliamson committed:
locale.c: Check for illegal input category

It is possible for Perl_setlocale() to be called with a category value
that is illegal, or is just unknown to us.  There is a function that
maps category numbers into our internal indexing scheme.  It maps any
unknown value to the same index, larger than any legal one.

Prior to this commit, there was no check that the input was valid; we
went off processing it as if it were valid.  That there was an element
in the tables that corresponds to this index prevented any illegal
memory acceses, but it is better to check early and return an error.

Since we compute the internal index for the category earlier in the
function, we can use the '_i' form of the internal calls rather than
calculating the index again.
